### PR TITLE
Create uk_ireland_accent_recognition.py

### PR DESCRIPTION
**Pull request by fbadine**
Adding a new example in the audio section that uses transfer learning to classify English (UK & Ireland) accents.
The example uses Yamnet to extract embeddings and then trains a model to classify the accents.

### CODE REVIEW

File: examples/audio/uk_ireland_accent_recognition.py [hunk 1]
```diff
@@ -0,0 +1,664 @@
+"""
+Title: UK & Ireland Accent Recognition Using Transfer Learning
+Author: [Fadi Badine](https://twitter.com/fadibadine)
+Date created: 16/04/2022
+Last modified: 16/04/2022
+Description: This notebook trains a model to classify UK & Ireland accents using feature extraction from Yamnet.
+"""
+
+"""
+# Prerequisits
```
Comment (fchollet) [inline examples/audio/uk_ireland_accent_recognition.py:10]: - All section titles should use `##` not `#`
- Please add an introduction section explaining what the example is about, what the reader will learn, what dataset you use, etc.

File: examples/audio/uk_ireland_accent_recognition.py [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+"""
+
+SEED = 1337
+EPOCHS = 100
+BATCH_SIZE = 64
+VALIDATION_RATIO = 0.1
+MODEL_NAME = "uk_irish_accent_recognition"
+
+# Location where the dataset will be downloaded.
+# By default (None), keras.utils.get_file will use /root/.keras/ as the CACHE_DIR
```
Comment (fchollet) [inline examples/audio/uk_ireland_accent_recognition.py:31]: It's `~/.keras` (it may not necessarily be `/root`, it depends on your system)

File: examples/audio/uk_ireland_accent_recognition.py [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+
+import matplotlib.pyplot as plt
+import seaborn as sns
+
+from scipy import stats
+from IPython.display import Audio
+
+print(f"TensorFlow: {tf.__version__}")
+print(f"TensorFlow Hub: {hub.__version__}")
+print(f"TensorFlow IO: {tfio.__version__}")
```
Comment (fchollet) [inline examples/audio/uk_ireland_accent_recognition.py:97]: No need for these statements; if the code requires specific versions you can just list them in the introduction

File: examples/audio/uk_ireland_accent_recognition.py [hunk 4]
```diff
@@ -0,0 +1,664 @@
...
+
+
+def d_prime(auc):
+    standard_normal = stats.norm()
+    d_prime = standard_normal.ppf(auc) * np.sqrt(2.0)
+    return d_prime
+
+
+def reset_random_seed(seed=1337):
```
Comment (fchollet) [inline examples/audio/uk_ireland_accent_recognition.py:106]: You can use `keras.utils.set_random_seed()` instead

File: examples/audio/uk_ireland_accent_recognition.py [hunk 5]
```diff
@@ -0,0 +1,664 @@
...
+"""
+### Evaluation
+"""
+
+train_loss, train_acc, train_auc = model.evaluate(train_ds)
+valid_loss, valid_acc, valid_auc = model.evaluate(valid_ds)
+
+print(
+    "train d-prime: {0:.3f}, validation d-prime: {1:.3f}".format(
+        d_prime(train_auc), d_prime(valid_auc)
```
Comment (fchollet) [inline examples/audio/uk_ireland_accent_recognition.py:465]: Define this function where it is used rather than upfront (and explain what it's for)

File: examples/audio/uk_ireland_accent_recognition.py [hunk 6]
```diff
@@ -0,0 +1,664 @@
...
+
+Since the dataset is quite unbalanced, we wil use class_weight during training.
+Getting the class weights is a little tricky because even though we know the number of audio files for each class, it does not represent the number of samples for that class since Yamnet transforms each audio file into multiple audio samples of 0.96 seconds each.
+So every audio file will be split into a number of samples that is proportional to its length.
+Therefore, to get those weights, we have to calculate the number of samples for each class after preprocessing through Yamnet.
+"""
+
+class_counts = tf.zeros(shape=(len(class_names),), dtype=tf.int32)
+
+for X, y in iter(train_ds):
```
Comment (fchollet) [inline examples/audio/uk_ireland_accent_recognition.py:392]: Use lowercase for variable names.

File: examples/audio/uk_ireland_accent_recognition.py [hunk 7]
```diff
@@ -0,0 +1,664 @@
...
+    x = keras.layers.Dense(384, activation="relu", name="dense_4")(x)
+    x = keras.layers.Dropout(0.2, name="dropout_4")(x)
+
+    outputs = keras.layers.Dense(len(class_names), activation="softmax", name="ouput")(
+        x
+    )
+
+    model = keras.Model(inputs=inputs, outputs=outputs, name="accent_recognition")
+
+    auc = keras.metrics.AUC(name="auc")
```
Comment (fchollet) [inline examples/audio/uk_ireland_accent_recognition.py:367]: You can inline this on the `metrics=` line

File: examples/audio/uk_ireland_accent_recognition.py [hunk 8]
```diff
@@ -0,0 +1,664 @@
...
+    # Using one-hot in order to use AUC
+    return (embeddings, tf.one_hot(labels, len(class_names)))
+
+
+def dataframe_to_dataset(dataframe, batch_size=64):
+    dataset = tf.data.Dataset.from_tensor_slices(
+        (dataframe["filename"], dataframe["label"])
+    )
+
+    dataset = dataset.map(lambda X, y: filepath_to_embeddings(X, y)).unbatch()
```
Comment (fchollet) [inline examples/audio/uk_ireland_accent_recognition.py:319]: Make sure to use `num_parallel_calls=tf.data.AUTOTUNE`

Comment (fchollet) [inline examples/audio/uk_ireland_accent_recognition.py:319]: Why `unbatch`?

File: examples/audio/uk_ireland_accent_recognition.py [hunk 9]
```diff
@@ -0,0 +1,664 @@
...
+## Results
+"""
+
+"""
+### TensorBoard
+"""
+
+"""
+```
+%reload_ext tensorboard
```
Comment (fchollet) [inline examples/audio/uk_ireland_accent_recognition.py:450]: Don't use TensorBoard for visualization since we cannot display this on the webpage. You could use matplotlib to display training/validation curves.

File: examples/audio/uk_ireland_accent_recognition.py [hunk 10]
```diff
@@ -0,0 +1,664 @@
...
+    )
+
+"""
+# Inference
+"""
+
+"""
+The below function `yamnet_class_names_from_csv` was copied and very slightly changed
+from [Yamnet
+Notebook](https://colab.research.google.com/github/tensorflow/hub/blob/master/examples/colab/yamnet.ipynb)
```
Comment (fchollet) [inline examples/audio/uk_ireland_accent_recognition.py:536]: Make sure to format all markdown links on a single line (see the guide: https://github.com/keras-team/keras-io/blob/master/contributor_guide.md)

File: examples/audio/uk_ireland_accent_recognition.py [hunk 11]
```diff
@@ -0,0 +1,664 @@
...
+filename = filename + ".wav"
+
+
+audio_wav, predictions, mel_spectrogram = filename_to_predictions(filename)
+
+infered_class = class_names[predictions.mean(axis=0).argmax()]
+print(f"The main accent is: {infered_class} English")
+
+"""
+## Results
```
Comment (fchollet) [inline examples/audio/uk_ireland_accent_recognition.py:615]: For each key section, make sure to have 1-2 sentences describing what's being done. It helps with the reading flow.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: examples/audio/uk_ireland_accent_recognition.py (hunk 8 above)
Comment (fbadine) [reply examples/audio/uk_ireland_accent_recognition.py:319]: Because in filepath_to_embeddings function, for each filename, we get a different number of embeddings as output depending on the audio file length. so for a filename, we get (N1, 1024) as output and (N2, 1024) for a different input file. An error occurs when I batch them. That's why I unbatch them and then batch them to batch_size

**Comment by fbadine** [comment]
Thanks @fchollet for your review and comments.
I did the requested changes.

**Review by fchollet — COMMENTED** [review]
Excellent! I applied some minor copyedits. Please address the last comment, and we should be ready to add the generated files.

File: examples/audio/uk_ireland_accent_recognition.py [hunk 12]
```diff
@@ -0,0 +1,700 @@
...
+yamnet_model = hub.load("https://tfhub.dev/google/yamnet/1")
+
+"""
+## Dataset
+The dataset used is the [Open-source Multi-speaker Corpora of the English Accents in the British Isles](https://openslr.org/83/) which consists of a total of 17,877 audio files.
+"""
+
+"""
+## Dataset Info
+@inproceedings{demirsahin-etal-2020-open,
```
[the remark this answers is not in the document]
Comment (fbadine) [reply examples/audio/uk_ireland_accent_recognition.py]: I removed the "Dataset Info" section and added more info about the dataset in the "Dataset" section

File: examples/audio/uk_ireland_accent_recognition.py [hunk 13]
```diff
@@ -0,0 +1,694 @@
+"""
+Title: UK & Ireland Accent Recognition Using Transfer Learning
```
Comment (fchollet) [inline examples/audio/uk_ireland_accent_recognition.py:2]: The title might be a bit too specific, perhaps "English speaker accent recognition using Transfer Learning" (note: don't capitalize every letter)

File: examples/audio/uk_ireland_accent_recognition.py [hunk 14]
```diff
@@ -0,0 +1,694 @@
...
+Author: [Fadi Badine](https://twitter.com/fadibadine)
+Date created: 16/04/2022
+Last modified: 16/04/2022
+Description: This notebook trains a model to classify UK & Ireland accents using feature extraction from Yamnet.
```
Comment (fchollet) [inline examples/audio/uk_ireland_accent_recognition.py:6]: "Training a model to classify UK & Ireland accents using feature extraction from Yamnet."

File: examples/audio/uk_ireland_accent_recognition.py [hunk 15]
```diff
@@ -0,0 +1,694 @@
+"""
+Title: UK & Ireland Accent Recognition Using Transfer Learning
+Author: [Fadi Badine](https://twitter.com/fadibadine)
+Date created: 16/04/2022
```
Comment (fchollet) [inline examples/audio/uk_ireland_accent_recognition.py:4]: Date order is inverted: it should be 2022/04/16

**Review by fchollet — COMMENTED** [review]
Thanks for the update! Some last comments about the header. Once addressed, please add the autogenerated files 👍

**Comment by fbadine** [comment]
Thanks for review @fchollet.
I have adjusted the header and generated the files

**Review by fchollet — APPROVED** [review]
Wonderful! Thank you for the great contribution 👍
